rfd/0144-advanced-okta-integration.md

```diff
@@ -0,0 +1,373 @@
+   list membership is evaluated (for example, at an access check).
+
+The idea of an _implicit_ access list extends this by removing the first
+condition. In an Implicit Access List, any user meeting the `membership_requires`
+conditions at evaluation time is implicitly considered a member of the list.
+
+While this is a useful property in its own right, it intersects with 
+the Okta sync service as it removes the need for any extra machinery to
+synchronize Access Lists to match the changing users coming from the
+upstream Okta system.
+
+Due to the Access List requirement that a User must meet the `membership_requires`
+conditions at the time membership is evaluated, most of the machinery 
+required to implement Implicit Access Lists is already in place.
+
+### Resource Changes
+
+To mark an Access List as *implicit*, we will add a `membership_evaluation` 
+field to the resource spec, with two possible values: 
+ * `explicit` - A user must be an explicitly-included member of the list to be 
+   considered a member. This is the default behaviour for backwards 
+   compatibility with existing lists.
+ * `implicit` - A user need only meet the `membership_requires` conditions at 
+   evaluation time to be considered a member of the list
+
+Example:
+
+```yaml
+version: v1
+kind: access_list
+metadata:
+  name: ea6cccbe-ceac-4776-8a89-4b1365fc03f5
+spec:
+  title: "Access List Title"
+  description: "A description of the Access List and its purpose"
+  membership_evaluation: "implicit"
+  # rest of the resource as per existing resource
+```
+
+### UX Changes
+
+The backends used by `tctl acl ls` and the Web UI will have to be updated to
+reflect that an Access List is Implicit, and to generate the list content at
+query time. 
+## Backwards Compatibility
+
+The activation of user sync and automatic SSO connector creation features will
+be controlled by new flags in the Okta plugin resource spec, with the features
+defaulting to *disabled*. This is done in order to prevent surprising behaviour
+in existing integrations when the new features are introduced in Teleport.
+
+This is not intended to be a generally available set of feature flags, and the 
+user will have no choice in the the enabled feature set. Any newly-created
+plugin resource will automatically have *all* of the Okta features supported
+by the creating version of Teleport enabled.
+## Staged release
+
+This is suggested structuring of the work required, such that each step delivers 
+useful features with minimal re-work in later stages.
+### Stage 1: Automation & Basic sync
+ * Automated creation of SSO SAML connector
+ * Limited Okta user sync
+   * Polling Only
+   * Teleport user created/destroyed in response to changes in Okta 
+   * **No** group memberships pulled from Okta (drastically reduces Okta API hit count)
+   * All users automatically get `access` and `requester` roles on creation. 
+ * Web UI treats okta users like local users (for example, Okta users may have traits 
```

Comment:
We should probably either include this in scope of stage 1 or fix shortly thereafter to avoid users messing things up accidentally.